sdk/search/azure-search-index/design.md

```diff
@@ -0,0 +1,483 @@
+# Azure Search Index Client Python SDK
+
+## Design
+
+Azure Cognitive Search is a fully managed cloud search service that provides a rich search experience to custom applications. The service provides a REST API with operations that create and manage indexes, load data, implement search features, execute queries, and handle results.
+
+This document descibes a Python SDK for Azure Search. The API provides a primary `SearchIndexClient` that implments all of the support REST operations of the service. Since the data in the index may not have a schema, documents and results are represented at simple (JSON) dicts. CRUD operations may be batched using an `IndexOperationBatch` class.
+
+*NOTE: This document borrows heavily from the existing work for the [Java Track 2 API for Azure Search](https://apiview.dev/Assemblies/Review/2abf935de65e4187964bea5b817872d6#com.azure.search.SearchIndexClient)*
+
+### Considerations
+
+#### POST/GET Variants
+
+Some of the service calls offer GET vs POST variants, with guidance similar to the following:
+
+> **When to use POST instead of GET**
+>
+>When you use HTTP GET to call Suggestions, the length of the request URL cannot exceed 8 KB. This length is usually enough for most applications. However, some applications produce very large queries, specifically when OData filter expressions are used. For these applications, HTTP POST is a better choice because it allows larger filters than GET. With POST, the number of clauses in a filter is the limiting factor, not the size of the raw filter string since the request size limit for POST is approximately 16 MB.
+
+In the generated Python API this is reflected in separate methods pairs such as `suggest_get` and `suggestpost`. This GET/POST dichotomy may be unfamiliar or confusing to many users, and does not seem like a useful distinction to directly expose in the user-facing API. Options incude:
+
+* Attempt to automatically choose the variant to use based on the length of the generated query string.
+* Always only use one of the variants (e.g. POST since it can accommodate longer queries).
+* Expose two explicit variants through a flag, using terminology that may be more semantically meaningful to users, e.g.
+  ```python
+  client.get_search_results(search_text="...", mode="simple")   # GET
+  client.get_search_results(search_text="...", mode="full")     # POST
+  ```
+
+***NOTE: Response from Bruce on this matter: "always use POST"***
+
+
+### `SearchApiKeyCredential` API
+
+The Azure Search service uses an HTTP header `"api-key"` to authenticate. This class abstracts the credential to provide room to support other methods of authentication that may become available in the future.
+
+```python
+azure.search.SearchApiKeyCredential(api_key)
+
+# read only property
+SearchAPIKeyCredential.api_key: str
```

Comment:
Will this also support a method like `update_key` that lets you roll the credential?

Reply:
I think it certainly could (it just hasn't been mentioned as a potential requirement yet until now)

Reply:
The ability to roll credentials is a requirement.

Reply:
We already have a management SDK in Python that can roll API keys.

Reply:
My understanding of the situation is that the mgmt API could be used to obtain the new credentials and then this requirement mentioned here is for long running python processes to be able to install those new credentials while still running. Is that correct @johanste ?

Reply:
FWIW, our Track 2 Java SDK has no such capability to roll an api-key in a running process (at least not that I'm aware of).

Reply:
Correct. The application would "push" the new key value into the credential. The credential/client would not reach out and get the new value.

Reply:
I've added `update_key` to be consistent with https://github.com/Azure/azure-sdk-for-python/pull/9406